### Building a predictive model by using probability of past records

In [130]:
import csv
import pandas as pd
import numpy as np
import glob

In [131]:
path =r'/media/romit/249E6A809E6A49FE/my_work/core_project/fol' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,index_col=None, header=0)
    list_.append(df)
frame= pd.concat(list_)
len(list_)

frame['winner_name'].replace(' ', '_',regex=True,inplace=True)
frame['loser_name'].replace(' ', '_',regex=True,inplace=True)

In [132]:
x=frame['winner_name'].unique().tolist()
v=frame['loser_name'].unique().tolist()
res_f=list(set(x+v))

In [133]:
frame['winner']=frame[['winner_name']]
frame=frame[['winner_name','loser_name','surface','round','winner','w_SvGms','l_SvGms']].copy()


In [134]:
dfs=pd.read_csv('serve_analysis.csv')
dfr=pd.read_csv('return_analysis.csv')
dfrnd=pd.read_csv('round_abilities.csv')
dfsrf=pd.read_csv('surface_abilities.csv')

In [135]:
dfrnd['Player_name'].replace(' ', '_',regex=True,inplace=True)
dfsrf['Player_name'].replace(' ', '_',regex=True,inplace=True)

In [136]:
x1=dfs['Player_name'].tolist()
x2=dfr['Player_name'].tolist()
x3=dfrnd['Player_name'].tolist()
x4=dfsrf['Player_name'].tolist()

x=set(x1) & set(x2)& set(x3) & set(x4)
len(x)

94

In [137]:
def player_prof(dfs,dfr,dfrnd,dfsrf,res_f):
    player=[]
    cols=['Player_name','Serve','Return','Surface','Round']
    pp=pd.DataFrame(columns=cols)
    
    for z in range(len(res_f)):
        array=[]
        array.append(res_f[z])
        serve=(dfs[dfs['Player_name']==res_f[z]]['f(i)']).sum()
        array.append(serve)
        
        ret=(dfr[dfr['Player_name']==res_f[z]]['g(i)']).sum()
        array.append(ret)
        
        r1=(dfrnd[dfrnd['Player_name']==res_f[z]]['R128']).sum()
        array.append(r1)
        
        r2=(dfrnd[dfrnd['Player_name']==res_f[z]]['R64']).sum()
        array.append(r2)
        
        r3=(dfrnd[dfrnd['Player_name']==res_f[z]]['R32']).sum()
        array.append(r3)
        
        r4=(dfrnd[dfrnd['Player_name']==res_f[z]]['R16']).sum()
        array.append(r4)
        
        QF=(dfrnd[dfrnd['Player_name']==res_f[z]]['QF']).sum()
        array.append(QF)
        
        SF=(dfrnd[dfrnd['Player_name']==res_f[z]]['SF']).sum()
        array.append(SF)
        
        F=(dfrnd[dfrnd['Player_name']==res_f[z]]['F']).sum()
        array.append(F)
        
        grass=(dfsrf[dfsrf['Player_name']==res_f[z]]['W_Grass']).sum()
        array.append(grass)
        
        clay=(dfsrf[dfsrf['Player_name']==res_f[z]]['W_Clay']).sum()
        array.append(clay)
        
        carpet=(dfsrf[dfsrf['Player_name']==res_f[z]]['W_Carpet']).sum()
        array.append(carpet)
    
        hard=(dfsrf[dfsrf['Player_name']==res_f[z]]['W_Hard']).sum()
        array.append(hard)
        
        player.append(array)
        
    return player

In [138]:
t=player_prof(dfs,dfr,dfrnd,dfsrf,res_f)


In [12]:
columns=['Player_name','serve','return','r1','r2','r3','r4','QF','SF','F','Grass','Clay','Carpet','Hard']
play_prof=pd.DataFrame(t,columns=columns)

In [13]:
play_prof=play_prof.dropna()

Player_name     serve    return        r1        r2  \
13              Gilles_Simon  0.596983  0.370257  0.518293  0.721311   
43            Michael_Berrer  0.693694  0.361502  0.325581  0.264706   
92             Yuichi_Sugita  0.597345  0.299046  0.500000  0.277778   
125             David_Ferrer  0.557681  0.346630  0.700441  0.717557   
126           Jonas_Bjorkman  0.575000  0.286638  0.540000  0.323077   
163            Gaston_Gaudio  0.536913  0.471775  0.646154  0.614286   
222         Julien_Benneteau  0.613497  0.338498  0.443299  0.526316   
226              Blaz_Kavcic  0.632353  0.441658  0.210526  0.350000   
242         Janko_Tipsarevic  0.558442  0.220543  0.516129  0.523810   
245           Michael_Llodra  0.614173  0.311223  0.444444  0.356322   
249           Thomas_Enqvist  0.633431  0.291636  0.571429  0.461538   
261           Ernests_Gulbis  0.631336  0.345145  0.469880  0.423529   
265           Martin_Verkerk  0.470000  0.245604  0.526316  0.500000   
289          Gustavo_Kuerten  0.619048  0.369797  0.612903  0.447368   
296              Yen_Hsun_Lu  0.697479  0.267472  0.355932  0.390244   
305              James_Blake  0.671642  0.411679  0.622642  0.628571   
310             Andy_Roddick  0.689560  0.300590  0.825806  0.780220   
329            Tomas_Berdych  0.661533  0.324394  0.665049  0.775862   
343   Guillermo_Garcia_Lopez  0.628571  0.418793  0.508621  0.390476   
355    Juan_Martin_Del_Potro  0.660441  0.337137  0.702703  0.724638   
367       Sebastien_Grosjean  0.573171  0.293204  0.682540  0.562500   
381            Florian_Mayer  0.578947  0.285023  0.540230  0.493827   
395            Alex_Corretja  0.584158  0.359465  0.424242  0.678571   
481           Arnaud_Clement  0.522222  0.376442  0.507463  0.458333   
488        Michal_Przysiezny  0.614035  0.099131  0.272727  0.111111   
489           Felix_Mantilla  0.554054  0.331413  0.463415  0.717949   
494          Jarkko_Nieminen  0.534426  0.310578  0.458333  0.481203   
508         Thomas_Johansson  0.713043  0.375648  0.543860  0.540984   
516            Roger_Federer  0.676349  0.376559  0.863830  0.922414   
545        Fernando_Verdasco  0.612267  0.299403  0.525140  0.618750   
...                      ...       ...       ...       ...       ...   
1405              Max_Mirnyi  0.525641  0.175975  0.483333  0.428571   
1408       Fernando_Gonzalez  0.597633  0.301531  0.654206  0.752577   
1429           Jurgen_Melzer  0.558824  0.230934  0.573643  0.436975   
1436   Philipp_Kohlschreiber  0.672566  0.352984  0.556250  0.588710   
1448            Gael_Monfils  0.657505  0.333748  0.679104  0.682243   
1462           Tommy_Robredo  0.626904  0.378969  0.518325  0.724138   
1465        Nicolas_Lapentti  0.382979  0.279705  0.428571  0.458333   
1492         Dmitry_Tursunov  0.670588  0.235525  0.487500  0.454545   
1498         Guillermo_Coria  0.588235  0.438346  0.559322  0.760000   
1525          Novak_Djokovic  0.652237  0.392951  0.875706  0.876543   
1557        Marcos_Baghdatis  0.621993  0.367761  0.547009  0.510204   
1562          Jerzy_Janowicz  0.500000  0.324055  0.466667  0.526316   
1563              Tommy_Haas  0.564202  0.320006  0.571429  0.721154   
1586      Yevgeny_Kafelnikov  0.607988  0.299725  0.533333  0.500000   
1604       Sergiy_Stakhovsky  0.621762  0.304957  0.322034  0.433962   
1606      Paul_Henri_Mathieu  0.619718  0.327899  0.433628  0.477064   
1671       Nikolay_Davydenko  0.703704  0.448387  0.568047  0.519608   
1700       Younes_El_Aynaoui  0.652174  0.234610  0.702703  0.447368   
1712       Ricardas_Berankis  0.570175  0.255128  0.500000  0.222222   
1746         Feliciano_Lopez  0.670718  0.294408  0.494048  0.579487   
1751     Juan_Carlos_Ferrero  0.648968  0.412923  0.567797  0.689320   
1753             Marat_Safin  0.687500  0.339685  0.595238  0.651685   
1766             Marin_Cilic  0.640511  0.317279  0.680851  0.689320   
1770            Igor_Andreev  0.431818  0.217234  0.45977

In [14]:
x=play_prof['Player_name'].tolist()
frame['count']=-3

for z in range(len(frame[:10])):
    count=0
    if frame.iloc[z]['winner_name'] in x:
        count=count+1
    if frame.iloc[z]['loser_name'] in x:
        count=count+1
    frame.iloc[z]['count']=count    
        
        

/home/romit/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


In [16]:
def count(frame,play_prof):
    array=[]
    for z in range(len(frame)):
        count=0
        if frame.iloc[z]['winner_name'] in x:
            count=count+1
        if frame.iloc[z]['loser_name'] in x:
            count=count+1
        #print count
        array.append(count) 

    return array

In [17]:
t=count(frame,play_prof)
l=pd.Series(t)

frame['count']=l

In [18]:
frame_m=frame[frame['count']==2]

In [19]:
frame_m=frame_m.drop(['count'],axis=1)
frame_m.dropna()

winner_name                  loser_name surface round  \
0                    Andy_Roddick           Nikolay_Davydenko    Hard   R32   
24                   Andy_Roddick              Michael_Llodra    Hard    QF   
32              Younes_El_Aynaoui                Andrei_Pavel    Hard   R32   
33              Rainer_Schuettler               Tommy_Robredo    Hard   R32   
47              Younes_El_Aynaoui         Juan_Carlos_Ferrero    Hard   R16   
48              Rainer_Schuettler               Roger_Federer    Hard   R16   
50                     Tim_Henman                  Max_Mirnyi    Hard   R16   
51                  Ivan_Ljubicic              Nicolas_Kiefer    Hard   R16   
55              Younes_El_Aynaoui           Rainer_Schuettler    Hard    QF   
56               Thomas_Johansson                  Tim_Henman    Hard    QF   
59              Younes_El_Aynaoui            Thomas_Johansson    Hard    SF   
72                  Tommy_Robredo          Paul_Henri_Mathieu    Hard   R32   
77             Yevgeny_Kafelnikov                  Max_Mirnyi    Hard   R32   
83             Sebastien_Grosjean               Tommy_Robredo    Hard   R16   
85             Yevgeny_Kafelnikov              Arnaud_Clement    Hard   R16   
88                 Thomas_Enqvist          Sebastien_Grosjean    Hard    QF   
89             Yevgeny_Kafelnikov                 Marc_Rosset    Hard    QF   
91                 Thomas_Enqvist          Yevgeny_Kafelnikov    Hard    SF   
96                 Martin_Verkerk               Ivan_Ljubicic    Hard   R64   
102                    Max_Mirnyi          Paul_Henri_Mathieu    Hard   R64   
108                Arnaud_Clement              Jonas_Bjorkman    Hard   R64   
123                Nicolas_Kiefer                  Max_Mirnyi    Hard   R32   
125                Felix_Mantilla           Younes_El_Aynaoui    Hard   R32   
134                Martin_Verkerk                  Tim_Henman    Hard   R16   
136            Sebastien_Grosjean              Nicolas_Kiefer    Hard   R16   
137                Felix_Mantilla                Andrei_Pavel    Hard   R16   
142                    Tommy_Haas          Sebastien_Grosjean    Hard    QF   
143                Felix_Mantilla              Arnaud_Clement    Hard    QF   
146                Felix_Mantilla           Rainer_Schuettler    Hard    SF   
155                 Ivan_Ljubicic              Martin_Verkerk    Hard   R32   
..                            ...                         ...     ...   ...   
329                   Mirza_Basic       Philipp_Kohlschreiber    Hard   R16   
334              Marcos_Baghdatis            Adrian_Mannarino    Hard   R16   
341                  Marius_Copil                 Robin_Haase    Hard   R32   
344                   Lukas_Lacko              Evgeny_Donskoy    Hard   R32   
345                 Jozef_Kovalik                  Radu_Albot    Hard   R32   
349                 Nicolas_Jarry              Roberto_Quiroz    Clay    RR   
353                Jose_Hernandez            Joao_Pedro_Sorgi    Hard    RR   
354               Thiago_Monteiro                 Roberto_Cid    Hard    RR   
356              Joao_Pedro_Sorgi                 Roberto_Cid    Hard    RR   
359                     Yibing_Wu          Jose_Rubin_Statham    Hard    RR   
361                    Aqeel_Khan             Seong_Chan_Hong   Grass    RR   
362         M_Abid_Ali_Khan_Akbar                Minjong_Park   Grass    RR   
364                 Dominic_Thiem            Dzmitry_Zhyrmont    Clay    RR   
365                 Dominic_Thiem                Ilya_Ivashka    Clay    RR   
367  Lloyd_George_Muirhead_Harris                 Edan_Leshem    Hard    RR   
375                Duilio_Beretta                Hugo_Dellien    Clay    RR   
392                  Skyler_Butts           Amir_Hossein_Badi    Hard    RR   
398           David_Agung_Susanto             Jeson_Patrombon    Hard    RR   
422                  Aljaz_Bedene              Hubert_Hurkacz    Hard   

## EVALUATION OF MODEL

### frame_m= DataSet


### play_prof=Evaluation model

In [115]:
import math

In [125]:
def model(play_prof,frame):
    a=[]
    f=float(play_prof['serve'].mean())
    g=float(play_prof['return'].mean())
    for z in range(len(frame)):
        x=frame.iloc[z]['winner_name']
        y=frame.iloc[z]['loser_name']
        
        x_sg=frame.iloc[z]['w_SvGms']
        y_sg=frame.iloc[z]['w_SvGms']
        
        srf=frame.iloc[z]['surface']
        rnd=frame.iloc[z]['round']
        
        
        
        winner=play_prof[play_prof['Player_name']==x]
        
        srv_w=winner['serve'].sum()
        ret_w=winner['return'].sum()
        
        if srf=='Grass':
            wp_s=winner['Grass'].sum()
        if srf=='Clay':
            wp_s=winner['Clay'].sum()
        if srf=='Hard':
            wp_s=winner['Hard'].sum()
        if srf=='Carpet':
            wp_s=winner['Carpet'].sum()
            
        if rnd=='R128':
            wp_r=winner['r1'].sum()
        if rnd=='R64':
             wp_r=winner['r2'].sum()
        if rnd=='R32':
             wp_r=winner['r3'].sum()
        if rnd=='R16':
            wp_r=winner['r4'].sum()
        if rnd=='QF':
            wp_r=winner['QF'].sum()
        if rnd=='SF':
             wp_r=winner['SF'].sum()
        if rnd=='F':
             wp_r=winner['F'].sum()
        
    
        
        looser=play_prof[play_prof['Player_name']==y]
        
        srv_l=looser['serve'].sum()
        ret_l=looser['return'].sum()
        
        if srf=='Grass':
            lp_s=looser['Grass'].sum()
        if srf=='Clay':
            lp_s=looser['Clay'].sum()
        if srf=='Hard':
            lp_s=looser['Hard'].sum()
        if srf=='Carpet':
            lp_s=looser['Carpet'].sum()
            
        if rnd=='R128':
            lp_r==looser['r1'].sum()
        if rnd=='R64':
            lp_r=looser['r2'].sum()
        if rnd=='R32':
            lp_r=looser['r3'].sum()
        if rnd=='R16':
            lp_r=looser['r4'].sum()
        if rnd=='QF':
            lp_r=looser['QF'].sum()
        if rnd=='SF':
            lp_r=looser['SF'].sum()
        if rnd=='F':
            lp_r=looser['F'].sum()
            
        
        f_xy=((srv_w-f)+(ret_w-g))
        f_yx=(srv_l-f)+(ret_l-g)
        
        '''ww=math.ceil(x_sg*(srv_w))
        ll=math.ceil(y_sg*(srv_l))
        
        
        t_x=(ww+(y_sg-ll))
        t_y=(ll+(x_sg-ww))'''
        
        
      
        
        if f_xy>=f_yx:
            a.append(y)
        else:
            a.append(x)
        
            
    return a
        

In [126]:
ser=model(play_prof,frame_m)

In [128]:
total=0
count=0
for z in range(len(frame_m)):
    if frame_m.iloc[z]['winner_name']==ser[z]:
        total=total+1
        count=count+1
       
    else:
        total=total+1
       
        
print count,total,float(count)/total
    

6724 8258 0.814240736256


## thus the accuracy of the model is 81%

##### the round and surface ability is yet to be incorporated into the model.